File: crates/unsnarl-root-query/src/parse_direction_token.rs
use crate::parse_error::ParseError;
use crate::root_query::Direction;

#[derive(Debug, PartialEq, Eq)]
pub struct ParsedDirectionToken {
    pub dir: Direction,
    pub level: Option<u32>,
}

pub fn parse_direction_token(text: &str) -> Result<ParsedDirectionToken, Vec<ParseError>> {
    let bytes = text.as_bytes();
    if bytes.len() < 2 || bytes[0] != b'+' {
        return Err(bad(text));
    }
    let dir = match bytes[1] {
        b'a' => Direction::After,
        b'b' => Direction::Before,
        b'c' => Direction::Context,
        _ => return Err(bad(text)),
    };
    let rest = &bytes[2..];
    if rest.is_empty() {
        return Ok(ParsedDirectionToken { dir, level: None });
    }
    if !rest.iter().all(|b| b.is_ascii_digit()) {
        return Err(bad(text));
    }
    let level: u32 = std::str::from_utf8(rest)
        .expect("ASCII digits are valid UTF-8")
        .parse()
        .map_err(|_| bad(text))?;
    Ok(ParsedDirectionToken {
        dir,
        level: Some(level),
    })
}

fn bad(text: &str) -> Vec<ParseError> {
    vec![ParseError {
        message: format!("unexpected direction token '{text}' (expected one of '+a', '+b', '+c')",),
    }]
}

#[cfg(test)]
#[path = "parse_direction_token_test.rs"]
mod parse_direction_token_test;

File: ts/src/root-query/parse-direction-token.ts
import type { ParseResult } from "./parse-error.js";

const DIR_TOKEN_RE = /^\+([abc])([0-9]*)$/;

type ParsedDirectionToken = Readonly<{
  dir: "a" | "b" | "c";
  level: number | null;
}>;

export function parseDirectionToken(
  text: string,
): ParseResult<ParsedDirectionToken> {
  const m = DIR_TOKEN_RE.exec(text);
  if (m === null) {
    return {
      ok: false,
      errors: [
        {
          message: `unexpected direction token '${text}' (expected one of '+a', '+b', '+c')`,
        },
      ],
    };
  }
  const dir = m[1] as "a" | "b" | "c";
  const levelStr = m[2] ?? "";
  const level = levelStr === "" ? null : Number.parseInt(levelStr, 10);
  return { ok: true, value: { dir, level } };
}
